state_machine_design.py

```diff
@@ -0,0 +1,355 @@
+        curr_time_seconds = hmsToSeconds(now.tm_hour, now.tm_min, now.tm_sec)
+
+        print("waking up at: " + str(curr_time_seconds))
+
+    def update(self, sm):
+        print("WAKE_UP update func")
+        # if active ensembles file not found or other reading with error,
+        # transition to ERROR state
+        if sm.err_code != NO_ERR:
+            sm.state = ERROR()
+        else:
+            # otherwise always
+            # transition to CHECK_TIME state
+            sm.state = CHECK_TIME()
+
+
+class CHECK_TIME(State):
+    '''
+    CHECK_TIME determines if the machine should skip an ensemble,
+    run an ensemble, or sleep until the next ensemble.
+    '''
+
+    def __init__(self):
+        self.check_time_ctrl = SKIP
+
+    def process(self, sm):
+        print("CHECK_TIME process func")
+        print("ens index: " + str(sm.ens_index))
+
+        if sm.ens_index < len(sm.ens_list):
+            # read time from ensemble and compare to current_time
+            now = time.localtime()
+            curr_time_seconds = hmsToSeconds(now.tm_hour, now.tm_min, now.tm_sec)
+
+            nearest_ens_time = sm.ens_list[sm.ens_index]["start_time"]
+
+            if nearest_ens_time < curr_time_seconds:
+                print("time past current ens, skipping")
+                self.check_time_ctrl = SKIP
+                sm.err_code = MISSED_ENS
+                # sm.err_msg = "Current ensemble past time, skipping ensemble "
+                # + sm.ens_list[sm.ens_index]["title"] + " at " + curr_time_seconds
+            elif nearest_ens_time == curr_time_seconds:
+                print("correct time for ensemble: " + sm.ens_list[sm.ens_index]["title"])
+                self.check_time_ctrl = NOW
+            else:
+                print("ensemble " + sm.ens_list[sm.ens_index]["title"] +
+                      " is still in future, go to sleep")
+                self.check_time_ctrl = WAIT
+        else:
+            print("index beyond last ens, resetting")
+            self.check_time_ctrl = RESET
+            sm.ens_index = 0
+
+
+    def update(self, sm):
+        print("CHECK_TIME update func")
+
+        # if current_time is passed current_ensemble time,
+        # transition to ITERATE
+        if self.check_time_ctrl == SKIP:
+            # sm.state = ITERATE() # switch with below line to disable error log on skip
+            sm.state = ERROR() # do we want to log this as error then recover?
+        # if current_time == current_ensemble time,
+        # transition to PERFORM_ENSEMBLE state
+        elif self.check_time_ctrl == NOW:
+            sm.state = PERFORM_ENSEMBLE()
+        # if current_time is less than current_ensemble time,
+        # transition to SLEEP state
+        elif self.check_time_ctrl == WAIT:
+            sm.state = SLEEP()
+        # if all ensembles are done, need to sleep til first one of next day
+        elif self.check_time_ctrl == RESET:
+            sm.state = SLEEP()
```

Comment:
I don't expect this to be something we use on deployment, but it'd be good to make this robust to the first ensemble needing to be performed immediately on reset. That is, if we've finished the last ensemble and `self.check_time_ctrl == RESET`, is there a case in which we would need to immediately transition to `PERFORM_ENSEMBLE` and can't afford the time to go to the `SLEEP` state?

Reply:
I think that the only time that this could happen is if the final ensemble finishes at the exact time of, or after the start time of the very first ensemble. That seemed unlikely to me so I didn't account for it. I've already written changes for the other issues so I'll think over this one tomorrow and make it able to handle that situation if you think it's a good idea to be able to. I don't think it'll be too difficult.

Reply:
Ok, I don't think it's a big deal but if you have time to include it I'd appreciate it.